lib/tools/system-calls.js

```diff
@@ -771,41 +772,50 @@ systemCallMethods.waitForDevice = async function (appDeviceReadyTimeout = 30) {
   });
 };
 
+/**
+ * Tries to run privileged shell command without root, otherwise elevates privileges and run them again
+ *
+ * @param {args} Array<string> - Shell arguments commands
+ */
+systemCallMethods.runPrivilegedShell = async function (args) {
+  let isRoot = false;
+  try {
+    await this.shell(args);
+  } catch (err) {
+    if (!err.message.includes('must be root')) {
+      throw err;
+    }
+    // this device needs adb to be running as root to run the specific command.
+    // so try to restart the daemon
+    log.debug(`Device requires adb to be running as root in order to run "adb shell ${quote(args)}". Restarting daemon`);
+    isRoot = await this.root();
+    await this.shell(args);
+  }
+  return isRoot;
+};
+
 /**
  * Reboot the current device and wait until it is completed.
  *
  * @param {number} retries [DEFAULT_ADB_REBOOT_RETRIES] - The maximum number of reboot retries.
  * @throws {Error} If the device failed to reboot and number of retries is exceeded.
  */
 systemCallMethods.reboot = async function (retries = DEFAULT_ADB_REBOOT_RETRIES) {
-  try {
-    try {
-      await this.shell(['stop']);
-    } catch (err) {
-      if (err.message.indexOf('must be root') === -1) {
-        throw err;
-      }
-      // this device needs adb to be running as root to stop.
-      // so try to restart the daemon
-      log.debug('Device requires adb to be running as root in order to reboot. Restarting daemon');
-      await this.root();
-      await this.shell(['stop']);
+  let isRoot = false;
```

Comment:
try/finally is still needed in case an exception is thrown somewhere in the middle of the method

Reply:
@mykola-mokhnach what should we set in the finally? We cannot unroot since this will raise exceptions in some cases.

Reply:
calling unroot is safe:

```
systemCallMethods.unroot = async function () {
  try {
    await exec(this.executable.path, ['unroot']);
    return true;
  } catch (err) {
    log.warn(`Unable to unroot adb daemon: '${err.message}'. Continuing`);
    return false;
  }
};
```

Reply:
It fails if the device goes offline for a couple of seconds.